File: api/offers.py
import hmac
import json
import os
import re
from datetime import datetime, timezone
from http.server import BaseHTTPRequestHandler
from typing import Any
from urllib.parse import parse_qs, urlparse


MAX_BODY_BYTES = 20_000


class ValidationError(Exception):
    pass


def _get_env(*names: str) -> str:
    for name in names:
        value = os.environ.get(name, "").strip()
        if value:
            return value
    return ""


def _get_supabase():
    url = _get_env("SUPABASE_URL")
    key = _get_env("SUPABASE_SERVICE_ROLE_KEY", "SUPABASE_KEY")
    if not url or not key:
        raise RuntimeError("Supabase configuration is missing")

    try:
        from supabase import create_client
    except ImportError as exc:
        raise RuntimeError("Supabase Python package is not installed") from exc

    return create_client(url, key)


def _clean_text(value: Any, max_len: int = 220) -> str:
    if value is None:
        return ""
    text = re.sub(r"\s+", " ", str(value)).strip()
    return text[:max_len]


def _to_bool(value: Any, fallback: bool = False) -> bool:
    if isinstance(value, bool):
        return value
    if value is None:
        return fallback
    return str(value).strip().lower() in {"1", "yes", "true", "active", "on"}


def _to_int(value: Any, fallback: int = 999) -> int:
    try:
        return int(value)
    except (TypeError, ValueError):
        return fallback


def _slugify(value: str) -> str:
    slug = re.sub(r"[^a-z0-9]+", "-", value.lower()).strip("-")
    return slug[:80] or f"offer-{int(datetime.now(timezone.utc).timestamp())}"


def _is_admin_request(headers) -> bool:
    configured_password = os.environ.get("ADMIN_PASSWORD", "").strip()
    if not configured_password:
        return False

    auth_header = headers.get("Authorization", "")
    provided = auth_header.removeprefix("Bearer ").strip()
    return hmac.compare_digest(provided, configured_password)


def _require_admin(headers):
    if not os.environ.get("ADMIN_PASSWORD", "").strip():
        raise PermissionError("Admin password is not configured.")
    if not _is_admin_request(headers):
        raise PermissionError("Invalid admin password.")


def _row_to_offer(row: dict[str, Any]) -> dict[str, Any]:
    return {
        "active": bool(row.get("active")),
        "priority": row.get("priority") or 999,
        "id": row.get("id") or "",
        "title": row.get("title") or "",
        "subtitle": row.get("subtitle") or "",
        "product": row.get("product") or "",
        "enquiryProduct": row.get("enquiry_product") or row.get("product") or "",
        "discount": row.get("discount") or "",
        "originalPrice": row.get("original_price") or "",
        "salePrice": row.get("sale_price") or "",
        "eventName": row.get("event_name") or "Store Offer",
        "badge": row.get("badge") or "",
        "endDate": row.get("end_date") or "",
        "image": row.get("image") or "",
        "highlight": bool(row.get("highlight")),
        "terms": row.get("terms") or "",
        "ctaText": row.get("cta_text") or "Enquire Now",
    }


def _offer_to_row(data: dict[str, Any]) -> dict[str, Any]:
    title = _clean_text(data.get("title"), 140)
    product = _clean_text(data.get("product"), 180)
    end_date = _clean_text(data.get("endDate") or data.get("end_date"), 80)

    if not title:
        raise ValidationError("Offer title is required.")
    if not product:
        raise ValidationError("Product name is required.")
    if not end_date:
        raise ValidationError("End date is required.")

    raw_id = _clean_text(data.get("id"), 90)
    offer_id = _slugify(raw_id or title)

    return {
        "id": offer_id,
        "active": _to_bool(data.get("active"), True),
        "priority": _to_int(data.get("priority"), 999),
        "title": title,
        "subtitle": _clean_text(data.get("subtitle"), 180),
        "product": product,
        "enquiry_product": _clean_text(data.get("enquiryProduct") or data.get("enquiry_product"), 180) or product,
        "discount": _clean_text(data.get("discount"), 40),
        "original_price": _clean_text(data.get("originalPrice") or data.get("original_price"), 60),
        "sale_price": _clean_text(data.get("salePrice") or data.get("sale_price"), 60),
        "event_name": _clean_text(data.get("eventName") or data.get("event_name"), 80) or "Store Offer",
        "badge": _clean_text(data.get("badge"), 50),
        "end_date": end_date,
        "image": _clean_text(data.get("image"), 500),
        "highlight": _to_bool(data.get("highlight"), False),
        "terms": _clean_text(data.get("terms"), 320),
        "cta_text": _clean_text(data.get("ctaText") or data.get("cta_text"), 50) or "Enquire Now",
        "updated_at": datetime.now(timezone.utc).isoformat(),
    }


def _sort_offers(offers: list[dict[str, Any]]) -> list[dict[str, Any]]:
    return sorted(
        offers,
        key=lambda offer: (
            not offer.get("highlight", False),
            offer.get("priority") or 999,
            (offer.get("title") or "").lower(),
        ),
    )


class handler(BaseHTTPRequestHandler):
    def _send_json(self, status: int, payload: dict):
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, DELETE, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type, Authorization")
        self.send_header("Cache-Control", "no-store")
        self.end_headers()
        self.wfile.write(json.dumps(payload).encode())

    def _read_json_body(self) -> dict:
        content_length = int(self.headers.get("Content-Length", 0))
        if content_length > MAX_BODY_BYTES:
            raise ValidationError("Request is too large.")

        raw_body = self.rfile.read(content_length)
        if not raw_body:
            return {}

        try:
            parsed = json.loads(raw_body)
        except json.JSONDecodeError:
            raise ValidationError("Invalid request body.")

        if not isinstance(parsed, dict):
            raise ValidationError("Invalid request body.")

        return parsed

    def _query(self):
        return parse_qs(urlparse(self.path).query)

    def do_GET(self):
        try:
            admin_mode = self._query().get("admin", [""])[0] == "1"
            if admin_mode:
                _require_admin(self.headers)

            result = _get_supabase().table("offers").select("*").execute()
            offers = [_row_to_offer(row) for row in (result.data or [])]
            if not admin_mode:
                offers = [offer for offer in offers if offer["active"]]

            self._send_json(200, {"offers": _sort_offers(offers)})
        except PermissionError as exc:
            self._send_json(401, {"error": str(exc)})
        except Exception as exc:
            print(f"Offers GET error: {exc}")
            self._send_json(500, {"error": "Could not load offers."})

    def do_POST(self):
        try:
            _require_admin(self.headers)
            offer_row = _offer_to_row(self._read_json_body())
            result = _get_supabase().table("offers").upsert(offer_row).execute()
            saved = result.data[0] if result.data else offer_row
            self._send_json(200, {"offer": _row_to_offer(saved)})
        except PermissionError as exc:
            self._send_json(401, {"error": str(exc)})
        except ValidationError as exc:
            self._send_json(422, {"error": str(exc)})
        except Exception as exc:
            print(f"Offers POST error: {exc}")
            self._send_json(500, {"error": "Could not save offer."})

    def do_DELETE(self):
        try:
            _require_admin(self.headers)
            offer_id = _clean_text(self._query().get("id", [""])[0], 90)
            if not offer_id:
                raise ValidationError("Offer ID is required.")

            _get_supabase().table("offers").delete().eq("id", _slugify(offer_id)).execute()
            self._send_json(200, {"message": "Offer deleted."})
        except PermissionError as exc:
            self._send_json(401, {"error": str(exc)})
        except ValidationError as exc:
            self._send_json(422, {"error": str(exc)})
        except Exception as exc:
            print(f"Offers DELETE error: {exc}")
            self._send_json(500, {"error": "Could not delete offer."})

    def do_OPTIONS(self):
        self._send_json(200, {})

File: api/offers.js
import crypto from 'node:crypto';

const MAX_BODY_BYTES = 20_000;
const SUPABASE_TIMEOUT_MS = 10_000;

class ValidationError extends Error {}

const getEnv = (...names) => {
  for (const name of names) {
    const value = (process.env[name] || '').trim();
    if (value) return value;
  }
  return '';
};

const getSupabaseConfig = () => {
  const url = getEnv('SUPABASE_URL').replace(/\/+$/, '');
  const key = getEnv('SUPABASE_SERVICE_ROLE_KEY', 'SUPABASE_KEY');

  if (!url || !key) {
    throw new Error('Supabase configuration is missing');
  }

  return { url, key };
};

const sendJson = (response, status, payload) => {
  response.statusCode = status;
  response.setHeader('Content-Type', 'application/json');
  response.setHeader('Access-Control-Allow-Origin', '*');
  response.setHeader('Access-Control-Allow-Methods', 'GET, POST, DELETE, OPTIONS');
  response.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');
  response.setHeader('Cache-Control', 'no-store');
  response.end(JSON.stringify(payload));
};

const cleanText = (value, maxLen = 220) => {
  if (value == null) return '';
  return String(value).replace(/\s+/g, ' ').trim().slice(0, maxLen);
};

const toBool = (value, fallback = false) => {
  if (typeof value === 'boolean') return value;
  if (value == null) return fallback;
  return ['1', 'yes', 'true', 'active', 'on'].includes(String(value).trim().toLowerCase());
};

const toInt = (value, fallback = 999) => {
  const parsed = Number.parseInt(value, 10);
  return Number.isFinite(parsed) ? parsed : fallback;
};

const slugify = (value) => {
  const slug = String(value || '')
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 80);

  return slug || `offer-${Math.floor(Date.now() / 1000)}`;
};

const timingSafeEqual = (left, right) => {
  const leftBuffer = Buffer.from(left);
  const rightBuffer = Buffer.from(right);
  return leftBuffer.length === rightBuffer.length && crypto.timingSafeEqual(leftBuffer, rightBuffer);
};

const isAdminRequest = (request) => {
  const configuredPassword = getEnv('ADMIN_PASSWORD');
  if (!configuredPassword) return false;

  const authHeader = request.headers.authorization || '';
  const provided = authHeader.replace(/^Bearer\s+/i, '').trim();
  return timingSafeEqual(provided, configuredPassword);
};

const requireAdmin = (request) => {
  if (!getEnv('ADMIN_PASSWORD')) {
    const error = new Error('Admin password is not configured.');
    error.status = 401;
    throw error;
  }

  if (!isAdminRequest(request)) {
    const error = new Error('Invalid admin password.');
    error.status = 401;
    throw error;
  }
};

const rowToOffer = (row) => ({
  active: Boolean(row.active),
  priority: row.priority || 999,
  id: row.id || '',
  title: row.title || '',
  subtitle: row.subtitle || '',
  product: row.product || '',
  enquiryProduct: row.enquiry_product || row.product || '',
  discount: row.discount || '',
  originalPrice: row.original_price || '',
  salePrice: row.sale_price || '',
  eventName: row.event_name || 'Store Offer',
  badge: row.badge || '',
  endDate: row.end_date || '',
  image: row.image || '',
  highlight: Boolean(row.highlight),
  terms: row.terms || '',
  ctaText: row.cta_text || 'Enquire Now',
});

const offerToRow = (data) => {
  const title = cleanText(data.title, 140);
  const product = cleanText(data.product, 180);
  const endDate = cleanText(data.endDate || data.end_date, 80);

  if (!title) throw new ValidationError('Offer title is required.');
  if (!product) throw new ValidationError('Product name is required.');
  if (!endDate) throw new ValidationError('End date is required.');

  const rawId = cleanText(data.id, 90);

  return {
    id: slugify(rawId || title),
    active: toBool(data.active, true),
    priority: toInt(data.priority, 999),
    title,
    subtitle: cleanText(data.subtitle, 180),
    product,
    enquiry_product: cleanText(data.enquiryProduct || data.enquiry_product, 180) || product,
    discount: cleanText(data.discount, 40),
    original_price: cleanText(data.originalPrice || data.original_price, 60),
    sale_price: cleanText(data.salePrice || data.sale_price, 60),
    event_name: cleanText(data.eventName || data.event_name, 80) || 'Store Offer',
    badge: cleanText(data.badge, 50),
    end_date: endDate,
    image: cleanText(data.image, 500),
    highlight: toBool(data.highlight, false),
    terms: cleanText(data.terms, 320),
    cta_text: cleanText(data.ctaText || data.cta_text, 50) || 'Enquire Now',
    updated_at: new Date().toISOString(),
  };
};

const sortOffers = (offers) => [...offers].sort((a, b) => {
  if (a.highlight !== b.highlight) return a.highlight ? -1 : 1;
  if (a.priority !== b.priority) return a.priority - b.priority;
  return a.title.localeCompare(b.title);
});

const getBody = async (request) => {
  if (request.body && typeof request.body === 'object') return request.body;
  if (typeof request.body === 'string') return request.body ? JSON.parse(request.body) : {};

  const chunks = [];
  let totalBytes = 0;

  for await (const chunk of request) {
    totalBytes += chunk.length;
    if (totalBytes > MAX_BODY_BYTES) {
      throw new ValidationError('Request is too large.');
    }
    chunks.push(chunk);
  }

  const rawBody = Buffer.concat(chunks).toString('utf8');
  return rawBody ? JSON.parse(rawBody) : {};
};

const supabaseRequest = async (method, table, query = {}, body = undefined, prefer = '') => {
  const { url, key } = getSupabaseConfig();
  const params = new URLSearchParams(query);
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), SUPABASE_TIMEOUT_MS);

  try {
    const result = await fetch(`${url}/rest/v1/${encodeURIComponent(table)}?${params}`, {
      method,
      headers: {
        apikey: key,
        Authorization: `Bearer ${key}`,
        'Content-Type': 'application/json',
        ...(prefer ? { Prefer: prefer } : {}),
      },
      body: body === undefined ? undefined : JSON.stringify(body),
      signal: controller.signal,
    });

    const text = await result.text();
    if (!result.ok) {
      throw new Error(`Supabase ${method} ${table} failed (${result.status}): ${text.slice(0, 500)}`);
    }

    return text ? JSON.parse(text) : [];
  } finally {
    clearTimeout(timeout);
  }
};

const handleGet = async (request, response) => {
  const requestUrl = new URL(request.url, 'https://local.vercel');
  const adminMode = requestUrl.searchParams.get('admin') === '1';

  if (adminMode) requireAdmin(request);

  const rows = await supabaseRequest('GET', 'offers', { select: '*' });
  let offers = rows.map(rowToOffer);

  if (!adminMode) {
    offers = offers.filter((offer) => offer.active);
  }

  sendJson(response, 200, { offers: sortOffers(offers) });
};

const handlePost = async (request, response) => {
  requireAdmin(request);

  const offerRow = offerToRow(await getBody(request));
  const result = await supabaseRequest(
    'POST',
    'offers',
    { on_conflict: 'id' },
    offerRow,
    'resolution=merge-duplicates,return=representation',
  );

  sendJson(response, 200, { offer: rowToOffer(result[0] || offerRow) });
};

const handleDelete = async (request, response) => {
  requireAdmin(request);

  const requestUrl = new URL(request.url, 'https://local.vercel');
  const offerId = cleanText(requestUrl.searchParams.get('id'), 90);
  if (!offerId) throw new ValidationError('Offer ID is required.');

  await supabaseRequest(
    'DELETE',
    'offers',
    { id: `eq.${slugify(offerId)}` },
    undefined,
    'return=minimal',
  );

  sendJson(response, 200, { message: 'Offer deleted.' });
};

export default async function handler(request, response) {
  try {
    if (request.method === 'OPTIONS') {
      sendJson(response, 200, {});
      return;
    }

    if (request.method === 'GET') {
      await handleGet(request, response);
      return;
    }

    if (request.method === 'POST') {
      await handlePost(request, response);
      return;
    }

    if (request.method === 'DELETE') {
      await handleDelete(request, response);
      return;
    }

    sendJson(response, 405, { error: 'Method not allowed.' });
  } catch (error) {
    if (error.status) {
      sendJson(response, error.status, { error: error.message });
      return;
    }

    if (error instanceof ValidationError) {
      sendJson(response, 422, { error: error.message });
      return;
    }

    console.error(`Offers ${request.method} error:`, error);
    const action = request.method === 'POST' ? 'save' : request.method === 'DELETE' ? 'delete' : 'load';
    sendJson(response, 500, { error: `Could not ${action} offers.` });
  }
}
